pd/src/components/validator_set.rs

```diff
@@ -83,61 +106,256 @@ impl ValidatorSet {
-            self.epoch = Some(new_epoch);
-            self.next_base_rate = None;
-            self.next_rates = None;
-            self.reward_notes = Vec::new();
-            self.supply_updates = BTreeMap::new();
+    /// Returns any changes to the validator set that occurred during
+    /// the course of the block.
+    pub fn block_changes(&self) -> Result<&BlockChanges> {
+        Ok(self
+            .block_changes
+            .as_ref()
+            .expect("block_changes called before end_block"))
+    }
+
+    /// Returns any changes to the validator set that occurred during
+    /// the course of the epoch.
+    pub fn epoch_changes(&self) -> Result<&EpochChanges> {
+        Ok(self
+            .epoch_changes
+            .as_ref()
+            .expect("epoch_changes called before end_epoch"))
+    }
+
+    // TODO: maybe the begin/end/commit flow should be a trait or something
+    pub fn begin_block(&mut self) {
+        self.block_changes = Some(Default::default());
+    }
+
+    pub fn begin_epoch(&mut self) {
+        self.epoch_changes = Some(Default::default());
+    }
+
+    /// Called during `commit_block` and will append database queries to save
+    /// the validator set, reset internal state for the next block, as well as
+    /// set `self.epoch` to the `new_epoch` value passed in.
+    pub async fn commit_block(
+        &mut self,
+        height: u64,
+        dbtx: &mut Transaction<'_, Postgres>,
+    ) -> Result<()> {
+        tracing::debug!(?height, "Committing block");
+        tracing::debug!("end height {}", self.epoch().end_height().value(),);
+
+        let block_changes = self.block_changes()?;
+
+        // Track the net change in delegations in this block.
+        let epoch_index = self.epoch().index;
+        for (identity_key, delegation_change) in &block_changes.delegation_changes {
+            query!(
+                "INSERT INTO delegation_changes VALUES ($1, $2, $3)",
+                identity_key.encode_to_vec(),
+                epoch_index as i64,
+                delegation_change
+            )
+            .execute(&mut *dbtx)
+            .await?;
         }
 
-        // TODO: split per-block and per-epoch state
-        // into separate structs wrapping the data
+        // Handle adding newly added validators with default rates
+        for v in &block_changes.new_validators {
+            query!(
+                "INSERT INTO validators (
+                    identity_key,
+                    consensus_key,
+                    sequence_number,
+                    name,
+                    website,
+                    description,
+                    voting_power,
+                    validator_state,
+                    unbonding_epoch
+                ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)",
+                v.validator.identity_key.encode_to_vec(),
+                v.validator.consensus_key.to_bytes(),
+                v.validator.sequence_number as i64,
```

Comment:
could there be an overflow issue here, where if the validator sets the sequence number to some value greater than max_i64, this will overflow to being negative, and cause replay to be possible for validator definitions?

Reply:
Good catch!